src/main/java/org/truffleruby/core/regexp/MatchDataNodes.java

```diff
@@ -203,16 +222,47 @@ public Object getIndex(DynamicObject matchData, int index, int length) {
             return createArray(store, length);
         }
 
+        @Specialization(guards = { "isRubySymbol(cachedIndex)", "name != null", "backRefs == 1",
+                "getRegexp(matchData) == regexp", "cachedIndex == index" })
+        public Object getIndexSymbolSingleMatch(DynamicObject matchData, DynamicObject index, NotProvided length,
+                @Cached("index") DynamicObject cachedIndex,
+                @Cached("getRegexp(matchData)") DynamicObject regexp,
+                @Cached("findNameEntry(regexp, index)") NameEntry name,
+                @Cached("numBackRefs(name)") int backRefs,
+                @Cached("backRefIndex(name)") int backRefIndex,
+                @Cached("createBinaryProfile()") ConditionProfile normalizedIndexProfile,
+                @Cached("createBinaryProfile()") ConditionProfile indexOutOfBoundsProfile,
+                @Cached("createBinaryProfile()") ConditionProfile hasValueProfile) {
+            return getIndex(matchData, backRefIndex, NotProvided.INSTANCE, normalizedIndexProfile, indexOutOfBoundsProfile, hasValueProfile);
+        }
+
+        @Specialization(guards = { "isRubySymbol(cachedIndex)", "name != null", "backRefs != 1",
+                "getRegexp(matchData) == regexp", "cachedIndex == index" })
+        public Object getIndexSymbol(DynamicObject matchData, DynamicObject index, NotProvided length,
+                @Cached("index") DynamicObject cachedIndex,
+                @Cached("getRegexp(matchData)") DynamicObject regexp,
+                @Cached("findNameEntry(regexp, index)") NameEntry name,
+                @Cached("numBackRefs(name)") int backRefs,
+                @Cached("createBinaryProfile()") ConditionProfile normalizedIndexProfile,
+                @Cached("createBinaryProfile()") ConditionProfile indexOutOfBoundsProfile,
+                @Cached("createBinaryProfile()") ConditionProfile hasValueProfile) {
+            final int i = getBackRef(matchData, regexp, name);
+
+            return getIndex(matchData, i, NotProvided.INSTANCE, normalizedIndexProfile, indexOutOfBoundsProfile, hasValueProfile);
```

Comment:
These two specializations could be merged, and just use `if backRefs == 1`.

Reply:
No, the previous specialisation is for cases where the name can match precisely one region number and is therefore stable, this specialisation is for cases where several regions have been associated with the name and the match data must be checked to see which one should be returned.

They are distinguished by the guard, either 'backRefs == 1` or `backRefs != 1`

Reply:
Yes, I know, but they could be merged in one specialization if `backRefIndex` just return some invalid value like -1 (it's not used anyway in that case) if backRefs != 1.
```java
@Specialization(guards = { "isRubySymbol(cachedIndex)", "name != null",
        "getRegexp(matchData) == regexp", "cachedIndex == index" })
public Object getIndexSymbolSingleMatch(VirtualFrame frame, DynamicObject matchData, DynamicObject index, NotProvided length,
        @Cached("index") DynamicObject cachedIndex,
        @Cached("getRegexp(matchData)") DynamicObject regexp,
        @Cached("findNameEntry(regexp, index)") NameEntry name,
        @Cached("numBackRefs(name)") int backRefs,
        @Cached("backRefIndex(name)") int backRefIndex) {
    if (backRefs == 1) {
        return executeGetIndex(frame, matchData, backRefIndex, NotProvided.INSTANCE);
    } else {
        final int i = getBackRef(matchData, regexp, name);
        return executeGetIndex(frame, matchData, i, NotProvided.INSTANCE);
    }
}
```